File: packages/exporter/src/infrastructure/assets/asset-registry.ts
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

export interface LatexAssetRegistry {
  assetRoot: string;
  styleFilePath: string;
  templatePath: string;
  logosDir: string;
  ucaldasLogoPath: string;
  aalieLogoPath: string;
}

function isValidAssetRoot(assetRoot: string): boolean {
  const stylePath = path.join(assetRoot, "aalie-report.sty");
  const templatePath = path.join(assetRoot, "templates", "main.template.tex");
  const ucaldasLogoPath = path.join(assetRoot, "logos", "ucaldas.pdf");
  const aalieLogoPath = path.join(assetRoot, "logos", "aalie.pdf");

  return (
    existsSync(stylePath) &&
    existsSync(templatePath) &&
    existsSync(ucaldasLogoPath) &&
    existsSync(aalieLogoPath)
  );
}

function collectCandidateRoots(): string[] {
  const candidates: string[] = [];
  const envRoot = process.env.AALIE_EXPORTER_ASSETS_DIR;
  if (envRoot) {
    candidates.push(path.resolve(envRoot));
  }

  const moduleDir = path.dirname(fileURLToPath(import.meta.url));
  candidates.push(path.resolve(moduleDir, "../../../assets/latex"));

  const cwd = process.cwd();
  candidates.push(path.resolve(cwd, "packages/exporter/assets/latex"));
  candidates.push(path.resolve(cwd, "../packages/exporter/assets/latex"));
  candidates.push(path.resolve(cwd, "../../packages/exporter/assets/latex"));

  let current = cwd;
  for (let depth = 0; depth < 8; depth += 1) {
    candidates.push(path.resolve(current, "packages/exporter/assets/latex"));
    const parent = path.dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }

  return Array.from(new Set(candidates));
}

export function resolveLatexAssetRegistry(): LatexAssetRegistry {
  const candidates = collectCandidateRoots();

  for (const candidate of candidates) {
    if (!isValidAssetRoot(candidate)) {
      continue;
    }

    return {
      assetRoot: candidate,
      styleFilePath: path.join(candidate, "aalie-report.sty"),
      templatePath: path.join(candidate, "templates", "main.template.tex"),
      logosDir: path.join(candidate, "logos"),
      ucaldasLogoPath: path.join(candidate, "logos", "ucaldas.pdf"),
      aalieLogoPath: path.join(candidate, "logos", "aalie.pdf"),
    };
  }

  throw new Error(
    [
      "Unable to resolve LaTeX assets for exporter.",
      "Expected files: aalie-report.sty, templates/main.template.tex, logos/ucaldas.pdf, logos/aalie.pdf.",
      `Checked candidates: ${candidates.join(", ")}`,
    ].join(" "),
  );
}

export function readLatexTemplate(registry: LatexAssetRegistry = resolveLatexAssetRegistry()): string {
  return readFileSync(registry.templatePath, "utf8");
}

File: apps/api/app/modules/export/asset_registry.py
"""
Asset registry resolution for backend-owned LaTeX assets.
"""

from __future__ import annotations

import os
from pathlib import Path

from .models import LatexAssetRegistry


def _is_valid_asset_root(root: Path) -> bool:
    return (
        (root / "aalie-report.sty").exists()
        and (root / "templates" / "main.template.tex").exists()
        and (root / "logos" / "ucaldas.pdf").exists()
        and (root / "logos" / "aalie.pdf").exists()
    )


def _candidate_roots() -> list[Path]:
    env_root = os.getenv("AALIE_EXPORTER_ASSETS_DIR")
    candidates: list[Path] = []
    if env_root:
        candidates.append(Path(env_root).expanduser().resolve())
    candidates.append((Path(__file__).resolve().parent / "assets" / "latex").resolve())
    unique: list[Path] = []
    seen: set[str] = set()
    for candidate in candidates:
        key = str(candidate)
        if key not in seen:
            seen.add(key)
            unique.append(candidate)
    return unique


def resolve_latex_asset_registry() -> LatexAssetRegistry:
    candidates = _candidate_roots()
    for candidate in candidates:
        if not _is_valid_asset_root(candidate):
            continue
        return LatexAssetRegistry(
            asset_root=str(candidate),
            style_file_path=str(candidate / "aalie-report.sty"),
            template_path=str(candidate / "templates" / "main.template.tex"),
            logos_dir=str(candidate / "logos"),
            ucaldas_logo_path=str(candidate / "logos" / "ucaldas.pdf"),
            aalie_logo_path=str(candidate / "logos" / "aalie.pdf"),
        )
    checked = ", ".join(str(path) for path in candidates)
    raise RuntimeError(
        "Unable to resolve LaTeX assets for exporter. "
        "Expected files: aalie-report.sty, templates/main.template.tex, "
        f"logos/ucaldas.pdf, logos/aalie.pdf. Checked candidates: {checked}"
    )


def read_latex_template(registry: LatexAssetRegistry | None = None) -> str:
    resolved = registry or resolve_latex_asset_registry()
    return Path(resolved.template_path).read_text(encoding="utf-8")
